controllers/reclaimspacecronjob_controller.go

```diff
@@ -0,0 +1,372 @@
+	return time.Now()
+}
+
+// constructRSJobForCronJob constructs reclaimspacejob.
+func (r *ReclaimSpaceCronJobReconciler) constructRSJobForCronJob(
+	rsCronJob *csiaddonsv1alpha1.ReclaimSpaceCronJob,
+	scheduledTime time.Time) (*csiaddonsv1alpha1.ReclaimSpaceJob, error) {
+	// We want job names for a given nominal start time to have a deterministic name to avoid the same job being created twice
+	name := fmt.Sprintf("%s-%d", rsCronJob.Name, scheduledTime.Unix())
+
+	job := &csiaddonsv1alpha1.ReclaimSpaceJob{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      make(map[string]string),
+			Annotations: make(map[string]string),
+			Name:        name,
+			Namespace:   rsCronJob.Namespace,
+		},
+		Spec: *rsCronJob.Spec.JobSpec.Spec.DeepCopy(),
+	}
+	for k, v := range rsCronJob.Spec.JobSpec.Annotations {
+		job.Annotations[k] = v
+	}
+	job.Annotations[scheduledTimeAnnotation] = scheduledTime.Format(time.RFC3339)
+	for k, v := range rsCronJob.Spec.JobSpec.Labels {
+		job.Labels[k] = v
+	}
+	if err := ctrl.SetControllerReference(rsCronJob, job, r.Scheme); err != nil {
+		return nil, err
+	}
+
+	return job, nil
+}
+
+// deleteOldJobs sorts given jobList by StartTime and deletes jobs older than given historyLimit.
+// Errors if any are only logged.
+func (r *ReclaimSpaceCronJobReconciler) deleteOldJobs(
+	ctx context.Context,
+	logger *logr.Logger,
+	jobsList []*csiaddonsv1alpha1.ReclaimSpaceJob,
+	historyLimit int32) {
+
+	sort.Slice(jobsList, func(i, j int) bool {
+		if jobsList[i].Status.StartTime == nil {
+			return jobsList[j].Status.StartTime != nil
+		}
+		return jobsList[i].Status.StartTime.Before(jobsList[j].Status.StartTime)
+	})
+
+	for i, job := range jobsList {
+		if int32(i) >= int32(len(jobsList))-historyLimit {
+			break
+		}
+		err := r.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationBackground))
+		if client.IgnoreNotFound(err) != nil {
+			logger.Error(err, "Failed to delete old job",
+				"reclaimSpaceJobName", job.Name,
+				"state", job.Status.Result)
+		} else {
+			logger.Info("Successfully deleted old job",
+				"reclaimSpaceJobName", job.Name,
+				"state", job.Status.Result)
+		}
+	}
+}
+
+// getScheduledTimeForRSJob extract the scheduled time from the annotation that
+// is added during job creation.
+func getScheduledTimeForRSJob(rsJob *csiaddonsv1alpha1.ReclaimSpaceJob) (*time.Time, error) {
+	timeRaw := rsJob.Annotations[scheduledTimeAnnotation]
+	if len(timeRaw) == 0 {
+		return nil, nil
+	}
+
+	timeParsed, err := time.Parse(time.RFC3339, timeRaw)
+	if err != nil {
+		return nil, err
+	}
+	return &timeParsed, nil
+}
+
+// TODO: add unit test for getNextSchedule
+// getNextSchedule returns lastMissed and next time after parsing the schedule.
+// This function returns error if there are more than 100 missed start times.
+func getNextSchedule(
+	rsCronJob *csiaddonsv1alpha1.ReclaimSpaceCronJob,
+	now time.Time) (lastMissed time.Time, next time.Time, err error) {
```

Comment:
`now time.Time) (lastMissed time.Time, next time.Time, err error) {` to `now time.Time) (ltime.Time,  time.Time,  error) {`